Caches: Refactored API

fulmeek opened the pull request:
Addressing issue #1037, this changes the Caches interface and current implementations to make it usable for different kind of caches:

- The methods `loadData`, `saveData`, `getTime` and `purgeCache` stay the same. You only have to ensure they consistently return `null` on non-existing data.
- **Removed**: `setPath($path)` and `setParameters(array $param)` were designed with file based caching in mind
- **New**:
  - **`setScope($scope)`: Set scope of the current cache** _(replaces `setPath`)_
    If `$scope` is an empty string, the cache is set to a global context. No absolute path anymore.
  - **`setKey($key)`: Set key to assign the current data** _(replaces `setParameters`)_
    Since `$key` can be anything, the cache implementation must ensure to assign the related data reliably; most commonly by serializing and hashing the key in an appropriate way.
- Cache implementations should check all requirements on construction (extensions, configuration). See `SQLiteCache` as an example.

Feel free to add comments and suggestions :)

fulmeek committed:
Caches: clean up existing CacheInterface

For consistency, functions should always return null on non-existing data.

fulmeek committed:
Caches: set scope

WordPressPluginUpdateBridge appears to have used its own cache instance in the past. Obviously not used anymore.

fulmeek committed:
Caches: set key to assign the current data

Since $key can be anything, the cache implementation must ensure to assign the related data reliably; most commonly by serializing and hashing the key in an appropriate way.

fulmeek committed:
[SQLiteCache] add configuration

Even though the default path for storage is perfectly fine, some people may want to use a different location.
This is an example how a cache implementation is responsible for its requirements.

fulmeek committed:
Caches: minor clean-up

logmanoriginal commented on caches/FileCache.php:
```diff
@@ -92,7 +94,13 @@ public function setParameters(array $param){
+		}
+
+		if(!is_dir($this->path)) {
+			if (mkdir($this->path, 0755, true) !== true) {
```
One minor change, not necessarily related to this PR. The permission 0755 should be 0644 because we don't want execute permission in the cache folder, right?

```suggestion
			if (mkdir($this->path, 0644, true) !== true) {
```

Haven't tested it though...

fulmeek replied:
Sorry, after testing, I had to revert it to 0755. The execute bit is required to change into that folder in order to read/write files in there. I didn't think about that earlier.

logmanoriginal reviewed (commented):
This looks good to me :+1: 

There are two things which I think should be worked on, however. One is mentioned below, the other is about protected functions. Maybe they should be made private to prevent confusion (because they are not supported by the interface and nobody should actually inherit these classes).

So, functions like `getPath`, `getCacheFile` and `getCacheName` in `FileCache` can be private and don't need to do additional parameter validation if this is already done by the calling function.

The same should be true for `getCacheKey` in `SQLiteCache`, right?

logmanoriginal committed:
Apply suggestions from code review

Co-Authored-By: fulmeek <36341513+fulmeek@users.noreply.github.com>

fulmeek committed:
[DisplayAction] use new cache functions

Done earlier, but forgotten to actually add these changes to the previous commits.

fulmeek committed:
Caches: make protected methods private + revert directory permissions

The previous code suggestions to set directory permissions to 644 had to reverted to 755 as the executable bit is required to change into that folder.

fulmeek commented:
Since it's been a while, any news?

logmanoriginal commented:
>Since it's been a while, any news?

Yes, I'm back :rofl: 

---

Seriously though, the changes look fine to me. Good job :+1: 
I'm gonna merge this but also need to point to #1000 because it'll break again.
